nx/lib/nx/binary_backend.ex

```diff
@@ -476,8 +476,101 @@ defmodule Nx.BinaryBackend do
   end
 
   @impl true
-  def dot(out, %{type: t1} = left, axes1, %{type: t2} = right, axes2) do
-    bin_zip_reduce(out, left, axes1, right, axes2, 0, fn lhs, rhs, acc ->
+  def dot(out, left, contract_axes1, [], right, contract_axes2, []) do
+    # dot/4 is directed to this specific clause so we can keep a more efficient implementation
+    # for non-batched dot products. See the clause below for batched dot products
+    dot4(out, left, contract_axes1, right, contract_axes2)
+  end
+
+  def dot(
+        out,
+        %{shape: left_shape, type: {_, left_size}, names: left_names} = left,
+        left_contract_axes,
+        left_batch_axes,
+        %{shape: right_shape, type: {_, right_size}, names: right_names} = right,
+        right_contract_axes,
+        right_batch_axes
+      ) do
+    left_binary = to_binary(left)
+    right_binary = to_binary(right)
+
+    left_batch_contract_axes = Nx.Shape.shift_axes(left_contract_axes, -length(left_batch_axes))
+
+    right_batch_contract_axes =
+      Nx.Shape.shift_axes(right_contract_axes, -length(right_batch_axes))
+
+    {left_batch_shape, left_batch_names} =
+      Nx.Shape.contract(left_shape, left_batch_axes, left_names, false)
+
+    {right_batch_shape, right_batch_names} =
+      Nx.Shape.contract(right_shape, right_batch_axes, right_names, false)
+
+    left_batch_size = Nx.size(left_batch_shape)
+    right_batch_size = Nx.size(right_batch_shape)
+
+    left_batch_count = Nx.Shape.batch_count(left_shape, left_batch_axes)
+    right_batch_count = Nx.Shape.batch_count(right_shape, right_batch_axes)
+
+    {batch_shape_out, _output_names} =
+      Nx.Shape.zip_reduce(
+        left_batch_shape,
+        left_batch_contract_axes,
+        left_batch_names,
+        right_batch_shape,
+        right_batch_contract_axes,
+        right_batch_names
+      )
+
+    left_count = Nx.size(left)
+    right_count = Nx.size(right)
+
+    count = max(left_count, right_count)
+
+    max_batch_count = max(left_batch_count, right_batch_count)
```

Comment:
Can we enforce the batch size is the same to skip this?

Reply:
Sure. This will simplify things!

Reply:
Actually, I think that without this we won't be able to have "dot(batched_tensor, normal_tensor)" or "dot(normal_tensor, batched_tensor)"

Reply:
Yes, this is the case

Reply:
I believe JAX (and XLA) requires the number of batch dimensions to be the same, so both tensors will have to be batched: 

```python
a = lax.iota(np.float32, 32)
a = lax.reshape(a, (2, 4, 4))
b = lax.iota(np.float32, 16)
b = lax.reshape(b, (4, 4))
lax.dot_general(a, b, (((2,), (0,)), ((0,), ())))
```

```
TypeError: dot_general requires equal numbers of lhs_batch and rhs_batch dimensions, got lhs_batch (0,) and rhs_batch ().
```